### PR TITLE
map identifiers to names

### PR DESCRIPTION
**Pull request by aleksdao**
Associated Issue: #741 

### Summary of Changes

* `mapIdentifiers` parses an expression via AST and maps identifiers to their minified values

### Test Plan

* wrote some assertions to test behavior of the `mapIdentifiers` function

### CODE REVIEW

**Comment by jasonLaster** [comment]
This is the first time we're using `babel-generator` I think it's a great usecase

File: packages/devtools-map-bindings/src/utils.js [hunk 1]
```diff
@@ -1,35 +1,31 @@
...
+      }
+    },
+  });
 
-  // Replace original name (which will be at the beginning of the expression)
-  // with found generated name.
-  const generatedName = foundScope.bindings[originalName];
-  const expressionWoOriginalName = expression.substring(originalName.length);
-  return `${generatedName}${expressionWoOriginalName}`;
+  return generate(ast, { concise: true, compact: true }).code.replace(";", "");
```
Comment (jasonLaster) [inline packages/devtools-map-bindings/src/utils.js]: what does this do: `.replace(";", "");`

File: packages/devtools-map-bindings/src/utils.js [hunk 2]
```diff
@@ -1,35 +1,31 @@
 // @flow
 
 import type { MappedScopeBindings } from "debugger-html";
+const { default: generate } = require("babel-generator");
```
Comment (jasonLaster) [inline packages/devtools-map-bindings/src/utils.js]: we usually do `const generate = require("babel-generator").default`

File: packages/devtools-map-bindings/src/utils.js (hunk 1 above)
Comment (aleksdao) [reply packages/devtools-map-bindings/src/utils.js]: `babel-generator` appears append a `;` on the generated output i.e. `"first*2"` becomes `"n*2;"`. To match the test assertions you provided, I removed the trailing `;`; thinking about this some more, there's probably a more standardized approach :)

Comment (yurydelendik) [reply packages/devtools-map-bindings/src/utils.js]: there may be a case when user typed ';' as part of the expression, e.g. `str.replace(";", "")`. it's better to use regexp there: `.replace(/;$/, "")`.

**Comment by yurydelendik** [comment]
Can you also add tests for:
- `window.first`
- `{ first: second }`
- `first;`
- `second.replace(";"g,"")`

**Comment by aleksdao** [comment]
@yurydelendik Of course! In your examples above, could you help me understand why we would not map many of those identifiers to their bindings? e.g. `window.first` => `window.first` (and not `window.n`) 

As the `mapIdentifiers` function is currently written, it picks up any `Identifier` node with `name === some_key_in_map_bindings` and replaces the original name with the mapped value. If the new test assertions indeed represent our desired output, I'll change the approach and update the PR. I'm happy to do it -- just wanted to get a confirmation. Thanks!

**Comment by jasonLaster** [comment]
@aleksdao i think the rationale for `window.first => window.first` is that if it is a property then we do not expect it to be different after the mapping.

The best way to check for this is to see if the identifier is a member expression

**Comment by aleksdao** [comment]
@jasonLaster Thanks! I'll continue to work off the assertions, as I think that will be the best way to ensure I'm handling every case. If you have additional inputs and expected outputs, let me know!

Both of the following are member expressions but will be handled differently.

`window.first` => `window.first`
`first[second]` => `n[u]`

**Comment by jasonLaster** [comment]
correct, by the way `foo.first => foo.first` should also work :) window is not special

**Comment by aleksdao** [comment]
@jasonLaster I'm having trouble differentiating between `first` => `n` and `first;` => `first;`. They seem to be represented by a nearly identical AST (the only difference being the `end` property in the `ExpressionStatement`.

`first`: [link](http://astexplorer.net/#/gist/3367194c1717033c14e5dc65a9d8404d/723a1d48b93d25994b56d98d9d987887793bc3ad)
`first;`:
 [link](http://astexplorer.net/#/gist/3367194c1717033c14e5dc65a9d8404d/f4dc678f6fa839a0b35de51b75b303f136c32fa6)

Do you have any suggestions?

**Comment by jasonLaster** [comment]
I’d use that, lets see if it’s parent is an `ExpressionStatement` :slightly_smiling_face:

**Comment by jasonLaster** [comment]
also join our [slack](https://devtools-html-slack.herokuapp.com) :wave:

**Comment by aleksdao** [comment]
@yurydelendik Can you confirm the expected outputs for

`first` -> ?
`first;` -> ?

besides the location, the ASTs for `first` and `first;` are identical. if we need to generate different outputs for them, it would be helpful to get a tip on how to differentiate between the two inputs

thanks!

**Comment by yurydelendik** [comment]
> ASTs for `first` and `first;` are identical.

If these are parsed as expressions the former is invalid one. If babel does not allow code to be parsed as expression, you can enclose that between `(` and `);` (and on the output strip these). As you can notice `(first;);` is invalid JS statement.

**Comment by jasonLaster** [comment]
i pushed a small tweak to the `;` case. and am asking the babel folks for a babely way to do it

File: packages/devtools-map-bindings/src/utils.js [hunk 3]
```diff
@@ -1,35 +1,33 @@
...
-  // JavaScript indetifier is a complex thing: it can contain '$', '\', ZWJ, any
-  // unicode letter or number. Simplifing that to latin character set for now.
-  // For simplicity, search only first characters of the expression to find an
-  // identifier/variable name.
-  const possibleVarNameRegex = /^([$\w]+)/;
-  const matchedVarName = possibleVarNameRegex.exec(expression);
-  if (!matchedVarName) {
-    return expression;
-  }
+  const ast = babylon.parse(expression, {
```
Comment (yurydelendik) [inline packages/devtools-map-bindings/src/utils.js]: babylon has the parseExpression method, which returns Expression tree (not Script)

**Comment by jasonLaster** [comment]
yep `parseExpression` is very strict though and needs a scope or other properties.

**Comment by yurydelendik** [comment]
> parseExpression is very strict though and needs a scope or other properties.

Correct, but that's what used by evaluateExpressions->evaluateExpression->hasSyntaxError

**Comment by jasonLaster** [comment]
Hmm, maybe that is what we should change then :)

**Comment by yurydelendik** [comment]
Okay, we need to properly define "expression" term. in JS, expression is defined as http://www.ecma-international.org/ecma-262/6.0/#sec-primary-expression . The console usage assumes that statements can be used (for, function, var, etc). Will we allow statements usage in the watch expressions? (I see it can be useful in the console though)

**Comment by jasonLaster** [comment]
@yurydelendik it's the same API, so yes... and conditional breakpoints benefit from statements too

**Comment by jasonLaster** [comment]
we really should let the server check for syntax errors too

**Comment by jasonLaster** [comment]
lets merge this so it's done and then relax the watch expression API

**Pull request merged**